### DISCUSSION TITLE
Generic Models in generic functions

### DISCUSSION THREAD

**Discussion by zoola969** [pydantic/pydantic#8855] [Feature Request]
I have some code. Let's assume that we have a generic function which receives generic data container and returns another generic data container with generic pydantic model depending on our initial generic variable. Instead of dummy `field=str(d)` we have sql or network request, which somehow removes our initial type and returns simple string.

So as we see according to mypy the code is correctly typed, but pydantic seems to not be able to extract current type of generic variable, and it parses the field value as a value of Any type. 

So is it possible to do what I want to do? 
  

```python
from decimal import Decimal
from typing import Any, Generic, TypeVar, reveal_type

from pydantic import BaseModel

_T = TypeVar("_T", bound=Any)


class GenericModel(BaseModel, Generic[_T]):
    field: _T


def generic_function(data: list[_T]) -> dict[_T, GenericModel[_T]]:
    return {d: GenericModel[_T](field=str(d)) for d in data}


def main():
    d1: list[int] = [1, 2, 3]
    d2: list[Decimal] = [Decimal("1.1"), Decimal("2.2"), Decimal("3.3")]
    r1 = generic_function(d1)
    r2 = generic_function(d2)
    reveal_type(r1)
    reveal_type(r2)
    print(type(r1[d1[0]].field))
    print(type(r2[d2[0]].field))
```

```bash
note: Revealed type is "builtins.dict[builtins.int, GenericModel[builtins.int]]"
note: Revealed type is "builtins.dict[_decimal.Decimal, GenericModel[_decimal.Decimal]]"
<class 'str'>
<class 'str'>
```

**Comment by GDYendell** [comment]
I am not sure I am understanding your example, but I think the confusion here is the static typing, not pydantic. At runtime those fields are `str`, because you are casting them to `str`. And in fact both mypy and pyright complain about this:

```
❯ mypy _pydantic.py 
_pydantic.py:210: error: Argument "field" to "GenericModel" has incompatible type "str"; expected "_T"  [arg-type]
```

```
❯ pyright _pydantic.py 
/workspaces/pvi/_pydantic.py
  /workspaces/pvi/_pydantic.py:210:39 - error: Argument of type "str" cannot be assigned to parameter "field" of type "_T@generic_function" in function "__init__"
    Type "str" cannot be assigned to type "_T@generic_function" (reportArgumentType)
```


If you print `r1` and `r2` you get 

```
{1: GenericModel(field='1'), 2: GenericModel(field='2'), 3: GenericModel(field='3')}
{Decimal('1.1'): GenericModel(field='1.1'), Decimal('2.2'): GenericModel(field='2.2'), Decimal('3.3'): GenericModel(field='3.3')}
```

If you take out the `str()` cast, you get

```
{1: GenericModel(field=1), 2: GenericModel(field=2), 3: GenericModel(field=3)}
{Decimal('1.1'): GenericModel(field=Decimal('1.1')), Decimal('2.2'): GenericModel(field=Decimal('2.2')), Decimal('3.3'): GenericModel(field=Decimal('3.3'))}
```

and mypy and pyright are both happy.

---

Maybe the simplification you describe is obscuring the problem you have with your real use case?

**Reply by zoola969** [reply]
Let me explain.

>At runtime those fields are str, because you are casting them to str. And in fact both mypy and pyright complain about this:
No, at runtime those fields are str only because at runtime `GenericModel` has `Any` as type of the filed `field`. For example if we explicitly indicate a type of the generic field, the we will correct parsing and result

```python
    print(type(GenericModel[Decimal](field="1.0").field))

    <class 'decimal.Decimal'>
```
My question is about how to (or is it possible) to know a concrete type of the generic at runtime and make a generic pydantic model parse data into this concrete type.
But after a couple of days reflecting on it, I have understood that it is most likely impossible, because we have to have classes with concrete types e.g. `GenericModel[Decimal]` on a compilation stage. 

BTW I came up with some workaround, may be it could be helpful in comprehending and solving the problem.

Here it is a little bit rewritten code to make the initial question more understandable (I hope so)
```python
from decimal import Decimal
from typing import Any, Generic, TypeVar, reveal_type

from pydantic import BaseModel, TypeAdapter

_T = TypeVar("_T", bound=Any)


class GenericModel(BaseModel, Generic[_T]):
    field: _T


def network_request(data: list[_T]) -> list[dict[str, str]]:
    return [{"field": str(d)} for d in data]


def generic_function(data: list[_T]) -> list[GenericModel[_T]]:
    type_adapter = TypeAdapter(list[GenericModel[_T]])
    response = network_request(data)
    return type_adapter.validate_python(response)


def main():
    d1: list[int] = [1, 2, 3]
    d2: list[Decimal] = [Decimal("1.1"), Decimal("2.2"), Decimal("3.3")]
    r1 = generic_function(d1)
    r2 = generic_function(d2)
    reveal_type(r1)
    reveal_type(r2)
    print(type(r1[0].field))
    print(type(r2[0].field))


if __name__ == "__main__":
    main()
```

And here it is the workaround:
```python
from decimal import Decimal
from typing import Any, Generic, TypeVar, overload, reveal_type

from pydantic import BaseModel, TypeAdapter

_T = TypeVar("_T", int, Decimal)


class GenericModel(BaseModel, Generic[_T]):
    field: _T


def network_request(data: list[_T]) -> list[dict[str, str]]:
    return [{"field": str(d)} for d in data]


def generic_function(data: list[_T]) -> list[GenericModel[_T]]:
    response = network_request(data)
    return parse_result(response, type(data[0]))


@overload
def parse_result(data: list[dict[str, str]], type_: type[int]) -> list[GenericModel[int]]: ...


@overload
def parse_result(data: list[dict[str, str]], type_: type[Decimal]) -> list[GenericModel[Decimal]]: ...


def parse_result(data: list[dict[str, str]], type_: type[_T]) -> list[GenericModel[Decimal]] | list[GenericModel[int]]:
    if type_ == int:
        type_adapter: TypeAdapter[list[GenericModel[int]] | list[GenericModel[Decimal]]] = TypeAdapter(
            list[GenericModel[int]]
        )
        return type_adapter.validate_python(data)
    type_adapter = TypeAdapter(list[GenericModel[Decimal]])
    return type_adapter.validate_python(data)


def main() -> None:
    d1: list[int] = [1, 2, 3]
    d2: list[Decimal] = [Decimal("1.1"), Decimal("2.2"), Decimal("3.3")]
    r1 = generic_function(d1)
    r2 = generic_function(d2)
    reveal_type(r1)
    reveal_type(r2)
    print(type(r1[0].field))
    print(type(r2[0].field))


if __name__ == "__main__":
    main()
```

As I said before, we need to create concrete generic models before the code starts executing. In this case both mypy and runtime types are correct and they are the same

**Reply by GDYendell** [reply]
Yes I agree. If you want pydantic to validate to specific types then you have to provide restrictions, as you have done in your second example by adding bounds to `_T`.

It will create the first model that matches during validation, so if it gets `"1"` as the serialsed form, and the field is `Any`, then it can validate with `field` as `str`. When `_T` is only allowed to be `int` or `Decimal`, then the model can't validate `"1"` as `str`, only as `int` or `Decimal`, and can only validate `"1.1"` as `Decimal`.

I don't believe pydantic pays any attention to the typing on functions models are returned from, if this is what you were hoping. So, pydantic doesn't know to validate the given values to `int` because that is the return type of `generic_function` when given a list of `int`. It only needs to be internally consistent for validation to succeed, so even though within the context of the function `_T` is `int`, when the model was built there was no such restriction.

**Comment by jonathanslenders** [comment]
Have a look at the decorator in this proposal which can help: https://github.com/python/typing/discussions/2199
